cupy/random/cupy_distributions.cu

```diff
@@ -0,0 +1,211 @@
+                X = pow(U, 1./shape);
+                if (X <= V) {
+                    return X;
+                }
+            } else {
+                Y = -log((1-U)/shape);
+                X = pow(1.0 - shape + shape*Y, 1./shape);
+                if (X <= (V + Y)) {
+                    return X;
+                }
+            }
+        }
+    } else {
+        b = shape - 1./3.;
+        c = 1./sqrt(9*b);
+        for (;;) {
+            do {
+                X = state->rk_normal();
+                V = 1.0 + c*X;
+            } while (V <= 0.0);
+            V = V*V*V;
+            U = state->rk_double();
+            if (U < 1.0 - 0.0331*(X*X)*(X*X)) return (b*V);
+            if (log(U) < 0.5*X*X + b*(1. - V + log(V))) return (b*V);
+        }
+    }
+}
+
+__device__ double rk_beta(rk_state* state, double a, double b) {
+    double Ga, Gb;
+    if ((a <= 1.0) && (b <= 1.0)) {
+        double U, V, X, Y;
+        /* Use Johnk's algorithm */
+        while (1) {
+            U = state->rk_double();
+            V = state->rk_double();
+            X = pow(U, 1.0/a);
+            Y = pow(V, 1.0/b);
+            if ((X + Y) <= 1.0) {
+                if (X +Y > 0) {
+                    return X / (X + Y);
+                } else {
+                    double logX = log(U) / a;
+                    double logY = log(V) / b;
+                    double logM = logX > logY ? logX : logY;
+                    logX -= logM;
+                    logY -= logM;
+                    return exp(logX - log(exp(logX) + exp(logY)));
+                }
+            }
+        }
+    } else {
+        Ga = rk_standard_gamma(state, a);
+        Gb = rk_standard_gamma(state, b);
+        return Ga/(Ga + Gb);
+    }
+}
+
+
+__device__ uint32_t rk_interval_32(rk_state* state, int mx, int mask) {
+    uint32_t sampled = state->rk_int() & mask;
+    while(sampled > mx)  {
+        sampled = state->rk_int() & mask;
+    }
+    return sampled;
+}
+
+__device__ uint64_t rk_interval_64(rk_state* state, uint64_t  mx, uint64_t mask) {
+    uint32_t hi= state->rk_int();
+    uint32_t lo= state->rk_int();
+    uint64_t sampled = (static_cast<uint64_t>(hi) << 32 | lo)  & mask;
+    while(sampled > mx)  {
+        hi= state->rk_int();
+        lo= state->rk_int();
+        sampled = (static_cast<uint64_t>(hi) << 32 | lo) & mask;
+    }
+    return sampled;
+}
+
+
+struct interval_32_functor {
+    template<typename... Args>
+    __device__ uint32_t operator () (Args&&... args) {
+        return rk_interval_32(args...);
+    }
+};
+
+struct interval_64_functor {
+    template<typename... Args>
+    __device__ uint64_t operator () (Args&&... args) {
+        return rk_interval_64(args...);
+    }
+};
+
+struct beta_functor {
+    template<typename... Args>
+    __device__ double operator () (Args&&... args) {
+        return rk_beta(args...);
+    }
+};
+
+// There are several errors when trying to do this a full template
+struct exponential_functor {
```

Comment:
Tried to create a single template for the functor, but since it is being used inside other templates there were a lot of conflicts with argument types and other stuff so I gave up.

Reply:
There is the issue that the memory pointer is in the device so we need to do stuff like `cudaMemcpyFromSymbol` which is not versatile at all